Реализации истории поиска.

AlexOtkidach opened the pull request:
Обновил ActivitySearch с логикой работы нового функционала истории поиска 
В разметке добавил кнопку "Очистить историю"

AlexOtkidach committed:
Добавил классы SearchHistory
Обновил ActivitySearch с логикой работы нового функционала истории поиска
В разметке добавил кнопку "Очистить историю"

AlexOtkidach committed:
Проверка Сохранения в Историю
Отображение Истории Поиска
Обновил разметку

НУЖНО ДОРАБОТАТЬ!!! GONE стории и самого поиска, решить проблему с UI .

AlexOtkidach committed:
Скорректирована работа плейсхолдеров
Ui элементов и истории поиска

DmitrijEgorow commented on app/src/main/java/com/example/plm2/SearchActivity.kt:
Добавляются не те треки


https://github.com/AlexOtkidach/plm2/assets/36837067/bc5f7de2-0800-4427-94b1-3b280195ec8c

AlexOtkidach replied:
Согласен, упустил этот момент, буду дорабатывать.

DmitrijEgorow commented on app/src/main/java/com/example/plm2/SearchActivity.kt:
Если список последних просмотренных треков пуст, то не отображаем заголовок и кнопку
![Screenshot_20231216_095132](https://github.com/AlexOtkidach/plm2/assets/36837067/636ef963-b6d4-484d-9265-e85d0ceb7671)

AlexOtkidach replied:
Понял

AlexOtkidach replied:
Поправил, теперь логика работы правильная и при очистке истории поиска все ее элементы включая заголовок "Очистить историю" скрываются.

DmitrijEgorow commented on app/src/main/java/com/example/plm2/SearchHistory.kt:
```diff
@@ -0,0 +1,44 @@
+
+class SearchHistory(private val sharedPreferences: SharedPreferences) {
+
+    companion object {
```
companion обычно идёт в конце

Подробнее описано в [документации](https://kotlinlang.org/docs/coding-conventions.html#class-layout)

AlexOtkidach replied:
Поправил, прочитал документацию теперь понятен порядок: 
Объявления свойств и блоки инициализаторов.
Вторичные конструкторы.
Объявления методов.
Companion object

DmitrijEgorow commented on app/src/main/java/com/example/plm2/SearchHistory.kt:
```diff
@@ -0,0 +1,44 @@
+
+
+
+
```
Много пустых строк, стоит отформатировать

AlexOtkidach replied:
Поправил, спасибо за замечание

DmitrijEgorow commented on app/src/main/java/com/example/plm2/SearchHistory.kt:
```diff
@@ -0,0 +1,44 @@
+
+    private fun saveSearchHistory(history: List<Track>) {
+        val json = Gson().toJson(history)
+        sharedPreferences.edit().putString("search_history", json).apply()
```
Уже определен ключ с таким значением

AlexOtkidach replied:
Заменил создание нового экземпляра gson на использование уже существующего экземпляра gson, определенного в companion object.

DmitrijEgorow commented on app/src/main/java/com/example/plm2/SearchActivity.kt:
```diff
@@ -211,89 +263,141 @@ class SearchActivity : AppCompatActivity() {
+    private fun updateSearchHistory(track: Track) {
+        Log.d("SearchActivity", "Updating search history with track: ${track.trackName}")
+        val historyList = getSearchHistory().toMutableList()
+        if (historyList.size >= 10) {
```
Магические числа – лучше выносить в константы с говорящими названиями

AlexOtkidach replied:
))) Спасибо за новый термин, начал гуглить, что такое "магическое число", оказывается так очень часто говорят в разработке)) Использовал константу MAX_HISTORY_SIZE которую как раз определил в companion object.

AlexOtkidach committed:
Исправил все ошибки и с добавлением треков в историю

AlexOtkidach commented:
Дмитрий, хотел бы сразу пояснить по кнопке "Очистки истории" поиска. Кнопка не динамическая (хотя такого требования нет, но согласно макету, я так подразумеваю именно этого и хотят), по той причине, она зафиксирована с отступом от нижнего края и края RecyclerView в 24 dp позволяет оставить ее всегда в видимости пользователя, ведь на небольшом экране от 6 и большом от 8 треков, они вместе с динамической кнопкой уходят за область видимости пользователя. Возможно, можно сделать кнопку динамической и одновременно с отступом от последнего элемента RecyclerView и при достижении края видимости экрана чтобы она фиксировалась с отступом. Но, я пока понятия не имею как это сделать))) и нагуглить не получилось (если вообще это возможно), могу сделать из 2-х описанных выше вариантов. Но повторюсь с точки зрения юзабилити,  кнопка в моей реализации всегда на виду, не перекрывается клавиатурой и не уходить за область видимости пользователя.

AlexOtkidach committed:
Интегрировал сохранение и восстановление состояния переключателя темы SwitchCompat с использованием SharedPreferences.

1. Теперь При каждом запуске SettingsActivity, приложение будет восстанавливать состояние переключателя темы на основе сохраненного значения в SharedPreferences и соответственно применять темную или светлую тему.

AlexOtkidach committed:
1. Исправил "Нечего не нашлось" -> "Ничего не нашлось"
2. Изменил плейсхолдер трека с офигенного динозавра) на тот, что в дизайне.
3. Тема приложения загружается из SharedPreferences при каждом запуске. Для этого использовал базовую активность.
